File: python/max_depth_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from typing import Deque


class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        queue = Deque()
        if root:
            queue.append(root)

        level = 0

        while len(queue) > 0:
            for i in range(len(queue)):
                node = queue.popleft()
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

            level += 1

        return level

File: cpp/maxDepthBinaryTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
#include <queue>
using namespace std;
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  int maxDepth(TreeNode *root) {
    queue<TreeNode *> queue;
    if (root) {
      queue.push(root);
    }

    int level = 0;

    while (queue.size() > 0) {
      int qLen = queue.size();
      level++;
      for (int i = 0; i < qLen; i++) {
        TreeNode *curr = queue.front();
        queue.pop();

        if (curr->left) {
          queue.push(curr->left);
        }
        if (curr->right) {
          queue.push(curr->right);
        }
      }
    }
    return level;
  }
};
